### PR TITLE
Updates

### PR DESCRIPTION
**Pull request by MRD2F**
- SignalObjectSelector: Fixed error in the PassLeptonVetoSelection method
- LeptonWeights : Updated prescale values

### CODE REVIEW

File: Analysis/src/EventInfo.cpp [hunk 1]
```diff
@@ -271,7 +271,9 @@ LorentzVector EventInfoBase::GetResonanceMomentum(bool useSVfit, bool addMET)
     Lock lock(*mutex);
     if(useSVfit && addMET)
         throw exception("Can't add MET and with SVfit applied.");
-    LorentzVector p4 = GetHiggsTTMomentum(useSVfit) + GetHiggsBB().GetMomentum();
+    LorentzVector p4 (0,0,0,0);
+    if(useSVfit && GetSVFitResults().has_valid_momentum)
+        p4 = GetHiggsTTMomentum(useSVfit) + GetHiggsBB().GetMomentum() ;
```
Comment (kandrosov) [inline Analysis/src/EventInfo.cpp]: wrong. should also consider cases where useSVfit=false.

File: Analysis/src/EventInfo.cpp [hunk 2]
```diff
@@ -293,25 +295,40 @@ const FatJetCandidate* EventInfoBase::SelectFatJet(double mass_cut, double delta
...
+                }
+            }
+            if((deltaR.at(0) < deltaR_subjet_cut && deltaR.at(3) < deltaR_subjet_cut)
+                    || (deltaR.at(1) < deltaR_subjet_cut && deltaR.at(2) < deltaR_subjet_cut))
+                return &fatJet;
+        }
+        else{
+            if(fatJet->p4().M() < mass_cut) continue;
+            std::vector<double> deltaR;
+            for(size_t n = 0; n < 2; ++n) {
```
Comment (kandrosov) [inline Analysis/src/EventInfo.cpp]: there should not be loop inside a loop. What is the purpose of the loop over n?

File: McCorrections/src/LeptonWeights.cpp [hunk 3]
```diff
@@ -185,35 +194,50 @@ double LeptonWeights::GetTriggerEfficiency(EventInfoBase& eventInfo, bool isData
...
-                if(period == Period::Run2017
-                        && !eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_unPrescaled)
-                        && eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_Prescaled_muTau))
-                    prescaled_weight = prescaled_weight_muTau;
+
+                if(period == Period::Run2016 &&
+                    !(eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_unPrescaled_2016) ||
+                        eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_Prescaled_IsoMu22_2016))
+                    && eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_Prescaled_IsoMu22_eta2p1_2016))
+                    prescaled_weight = prescaled_weight_IsoMu22_eta2p1_2016;
```
Comment (kandrosov) [inline McCorrections/src/LeptonWeights.cpp]: for the first 2016 prescale you should check only unprescaled, right?

File: McCorrections/src/LeptonWeights.cpp [hunk 4]
```diff
@@ -185,35 +194,50 @@ double LeptonWeights::GetTriggerEfficiency(EventInfoBase& eventInfo, bool isData
...
-        static const std::vector<std::string> triggerPaths_unPrescaled = {
+        static const std::vector<std::string> triggerPaths_unPrescaled_2017 = {
             "HLT_IsoMu27_v"};
         if(period == Period::Run2017
-                && !eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_unPrescaled)
-                && eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_Prescaled_muTau))
-            prescaled_weight = prescaled_weight_muTau;
+                && !eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_unPrescaled_2017)
+                && eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_Prescaled_muTau_2017))
+            prescaled_weight = prescaled_weight_muTau_2017;
```
Comment (kandrosov) [inline McCorrections/src/LeptonWeights.cpp]: add a comment for the future that for muMu for period Run2016 prescales should also be applied if all single mu triggers are used.

File: Analysis/src/EventInfo.cpp (hunk 1 above)
Comment (kandrosov) [reply Analysis/src/EventInfo.cpp]: still wrong.

File: McCorrections/src/LeptonWeights.cpp [hunk 5]
```diff
@@ -200,13 +200,20 @@ double LeptonWeights::GetTriggerEfficiency(EventInfoBase& eventInfo, bool isData
                 static const std::vector<std::string> triggerPaths_unPrescaled_2017 = {
                     "HLT_IsoMu27_v", "HLT_IsoMu20_eta2p1_LooseChargedIsoPFTau27_eta2p1_CrossL1_v" };
 
+
                 if(period == Period::Run2016 &&
+                    eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_unPrescaled_2016) &&
+                    (eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_Prescaled_IsoMu22_2016)
+                    || eventInfo.GetTriggerResults().AnyAcceptAndMatch(triggerPaths_Prescaled_IsoMu22_eta2p1_2016)))
+                    prescaled_weight = 1;
```
Comment (kandrosov) [inline McCorrections/src/LeptonWeights.cpp]: 1 is set by default this if is useless

File: McCorrections/src/LeptonWeights.cpp (hunk 3 above)
Comment (kandrosov) [reply McCorrections/src/LeptonWeights.cpp]: still not corrected

File: Analysis/src/EventInfo.cpp (hunk 1 above)
Comment (kandrosov) [reply Analysis/src/EventInfo.cpp]: wrong. what if svfit has not converged?

Comment (MRD2F) [reply Analysis/src/EventInfo.cpp]: There are GetSVFitResults().has_valid_momentum & useSVfit=false. You said that should be calculated also when is false. After, only requiring  valid momentum is wrong, that's what I understood that checked if svfit has not converged, and for that event the p4 remain 0. What is wrong with:

   ```
 if(GetSVFitResults().has_valid_momentum)
        p4 = GetHiggsTTMomentum(useSVfit) + GetHiggsBB().GetMomentum()
```
?
In the case that is not valid it remains p4(0,0,0,0)

Comment (kandrosov) [reply Analysis/src/EventInfo.cpp]: > What is wrong with:
> ```
> if(GetSVFitResults().has_valid_momentum)
>      p4 = GetHiggsTTMomentum(useSVfit) + GetHiggsBB().GetMomentum()
> ```
> ?

if `useSVfit=false`, `p4` shouldn't depend on whatever SVFit has valid momentum or not.

**Pull request merged**